### PR TITLE
fix: return correct locale in root 404 and 500 page with i18n

### LINKED ISSUE #12356: Astro.currentLocale not working correctly in root 404.astro with i18n
### Astro Info

```block
Astro                    v4.16.8
Node                     v22.9.0
System                   macOS (arm64)
Package Manager          npm
Output                   hybrid
Adapter                  @astrojs/vercel/serverless
Integrations             astro-icon
                         @sanity/astro
                         @astrojs/react
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When using a 404.astro page at the root level (pages/404.astro), Astro.currentLocale always returns the defaultLocale even when visiting a different locale URL path.

Example:
- URL: /de/nonexistent-page
- Astro.currentLocale returns 'fr' (defaultLocale) instead of 'de'

Console log shows:
{ currentLocale: 'fr', url: '/de/fasf', params: {} }

This works correctly in some projects but not others with similar configurations.

My config:
```astro.config.mjs
i18n: {
   defaultLocale: "fr",
   locales: ["fr", "de"],
   routing: {
      prefixDefaultLocale: true,
      redirectToDefaultLocale: false,
   }
}
```
### What's the expected result?

Astro.currentLocale should return 'de' when visiting /de/* paths 

### Link to Minimal Reproducible Example

-

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by apatel369**
## Changes
This PR fixes the issue where `Astro.currentLocale` was not correctly returning the locale for custom 404 and 500 pages in projects using i18n. 

Closes #12356 

## Testing

Manually verified in dev server and SSR.

Added Dev and SSR tests custom for vertifying locale in custom 404 pages.

## Docs
It is not needed as it is a bug fix

### CODE REVIEW

File: packages/astro/src/core/render-context.ts [hunk 1]
```diff
@@ -541,11 +541,9 @@ export class RenderContext {
 		}
 
 		let computedLocale;
-		if (routeData.pathname) {
-			computedLocale = computeCurrentLocale(routeData.pathname, locales, defaultLocale);
-		} else {
-			computedLocale = computeCurrentLocale(url.pathname, locales, defaultLocale);
-		}
+		const pathname =
+			routeData.pathname && routeData.pathname !== '/404' ? routeData.pathname : url.pathname;
```
[the remark this answers is not in the document]
Comment (apatel369) [reply packages/astro/src/core/render-context.ts:545]: Thanks. I will make those changes.

[the remark this answers is not in the document]
Comment (apatel369) [reply packages/astro/src/core/render-context.ts]: Added requested changes. PR is ready for re-review.

[the remark this answers is not in the document]
Comment (apatel369) [reply packages/astro/src/core/render-context.ts]: @ematipico PTAL!

**Review by ematipico — APPROVED** [review]

**Comment by ematipico** [comment]
The test failing seems to be a fluke, and it's not related to your changes. Thank you for the fix!

**Review by ematipico — APPROVED** [review]

**Pull request merged**